lfs/diff_index_scanner.go

```diff
@@ -0,0 +1,191 @@
+	"bufio"
+	"strconv"
+	"strings"
+
+	"github.com/git-lfs/git-lfs/errors"
+)
+
+// Status represents the status of a file that appears in the output of `git
+// diff-index`.
+//
+// More information about each of its valid instances can be found:
+// https://git-scm.com/docs/git-diff-index
+type Status rune
+
+const (
+	StatusAddition     Status = 'A'
+	StatusCopy         Status = 'C'
+	StatusDeletion     Status = 'D'
+	StatusModification Status = 'M'
+	StatusRename       Status = 'R'
+	StatusTypeChange   Status = 'T'
+	StatusUnmerged     Status = 'U'
+	StatusUnknown      Status = 'X'
+)
+
+// String implements fmt.Stringer by returning a huamn-readable name for each
+// status.
+func (s Status) String() string {
+	switch s {
+	case StatusAddition:
+		return "addition"
+	case StatusCopy:
+		return "copy"
+	case StatusDeletion:
+		return "deletion"
+	case StatusModification:
+		return "modification"
+	case StatusRename:
+		return "rename"
+	case StatusTypeChange:
+		return "change"
+	case StatusUnmerged:
+		return "unmerged"
+	case StatusUnknown:
+		return "unknown"
+	}
+	return "<unknown>"
+}
+
+// DiffIndexEntry holds information about a single item in the results of a `git
+// diff-index` command.
+type DiffIndexEntry struct {
+	// SrcMode is the file mode of the "src" file, stored as a string-based
+	// octal.
+	SrcMode string
+	// DstMode is the file mode of the "dst" file, stored as a string-based
+	// octal.
+	DstMode string
+	// SrcSha is the Git blob ID of the "src" file.
+	SrcSha string
+	// DstSha is the Git blob ID of the "dst" file.
+	DstSha string
+	// Status is the status of the file in the index.
+	Status Status
+	// StatusScore is the optional "score" assosicated with a particular
+	// status.
+	StatusScore int
+	// SrcName is the name of the file in it's "src" state as it appears in
+	// the index.
+	SrcName string
+	// DstName is the name of the file in it's "dst" state as it appears in
+	// the index.
+	DstName string
+}
+
+// DiffIndexScanner scans the output of the `git diff-index` command.
+type DiffIndexScanner struct {
+	// next is the next entry scanned by the Scanner.
+	next *DiffIndexEntry
+	// err is any error that the Scanner encountered while scanning.
+	err error
+
+	// from is the underlying scanner, scanning the `git diff-index`
+	// command's stdout.
+	from *bufio.Scanner
+}
+
+// NewDiffIndexScanner initializes a new `DiffIndexScanner` scanning at the
+// given ref, "ref".
+//
+// If "cache" is given, the DiffIndexScanner will scan for differences between
+// the given ref and the index. If "cache" is _not_ given, DiffIndexScanner will
+// scan for differences between the given ref and the currently checked out
+// tree.
+//
+// If any error was encountered in starting the command or closing its `stdin`,
+// that error will be returned immediately. Otherwise, a `*DiffIndexScanner`
+// will be returned with a `nil` error.
+func NewDiffIndexScanner(ref string, cached bool) (*DiffIndexScanner, *wrappedCmd, error) {
```

Comment:
Last thing: the old `ScanIndex()` function had a `cmd.Wait()` after the `bufio.NewScanner(cmd.Stdout).Scan()` loop. I don't think that's necessary, since the cmd [will have already closed `exec/cmd.Stdout`][1] by the time that `Wait()` will have returned.

[1]: https://github.com/golang/go/blob/go1.8/src/os/exec/exec.go#L448